qutebrowser/misc/mhtml.py

```diff
@@ -0,0 +1,411 @@
+        """
+        msg = MIMEMultipart("related", self.BOUNDARY)
+
+        root = self._create_root_file()
+        msg.attach(root)
+
+        for file_data in self._files.values():
+            msg.attach(self._create_file(file_data))
+
+        gen = BytesGenerator(fp, policy=MHTMLPolicy)
+        gen.flatten(msg)
+
+    def _create_root_file(self):
+        """Return the root document as MIMEMultipart."""
+        root_file = _File(
+            content=self.root_content, content_type=self.content_type,
+            content_location=self.content_location, transfer_encoding=E_QUOPRI,
+        )
+        return self._create_file(root_file)
+
+    def _create_file(self, f):
+        """Return the single given file as MIMEMultipart."""
+        msg = MIMEMultipart()
+        msg["Content-Location"] = f.content_location
+        # Get rid of the default type multipart/mixed
+        del msg["Content-Type"]
+        if f.content_type:
+            msg.set_type(f.content_type)
+        encoding_name, encoding_func = f.transfer_encoding
+        if encoding_name:
+            msg["Content-Transfer-Encoding"] = encoding_name
+        msg.set_payload(encoding_func(f.content).decode("ascii"))
+        return msg
+
+
+class _Downloader(object):
+
+    """A class to download whole websites.
+
+    Attributes:
+        web_view: The QWebView which contains the website that will be saved.
+        dest: Destination filename.
+        writer: The MHTMLWriter object which is used to save the page.
+        loaded_urls: A set of QUrls of finished asset downloads.
+        pending_downloads: A set of unfinished (url, DownloadItem) tuples.
+        _finished: A flag indicating if the file has already been written.
+    """
+
+    def __init__(self, web_view, dest):
+        self.web_view = web_view
+        self.dest = dest
+        self.writer = MHTMLWriter()
+        self.loaded_urls = {web_view.url()}
+        self.pending_downloads = set()
+        self._finished = False
+
+    def run(self):
+        """Download and save the page.
+
+        The object must not be reused, you should create a new one if
+        you want to download another page.
+        """
+        web_url = self.web_view.url()
+        web_frame = self.web_view.page().mainFrame()
+
+        self.writer.root_content = web_frame.toHtml().encode("utf-8")
+        self.writer.content_location = web_url.toString()
+        # I've found no way of getting the content type of a QWebView, but
+        # since we're using .toHtml, it's probably safe to say that the
+        # content-type is HTML
+        self.writer.content_type = 'text/html; charset="UTF-8"'
+        # Currently only downloading <link> (stylesheets), <script>
+        # (javascript) and <img> (image) elements.
+        elements = (web_frame.findAllElements("link") +
+                    web_frame.findAllElements("script") +
+                    web_frame.findAllElements("img"))
+
+        for element in elements:
```

Comment:
Please wrap those in a `WebElementWrapper` (`browser/webelem.py`) - then there will be some checks done for you, and you can use dict-like access for attributes, e.g. `element['src']`.